azkaban-common/src/main/java/azkaban/executor/ExecutorManager.java

```diff
@@ -1866,29 +1865,67 @@ private void processQueuedFlows(final long activeExecutorsRefreshWindow,
-        final ExecutableFlow exflow, final Set<Executor> availableExecutors)
+        final ExecutableFlow exflow)
         throws ExecutorManagerException {
+      final Set<Executor> remainingExecutors = new HashSet<>(ExecutorManager.this.activeExecutors);
       synchronized (exflow) {
-        final Executor selectedExecutor = selectExecutor(exflow, availableExecutors);
-        if (selectedExecutor != null) {
-          try {
-            dispatch(reference, exflow, selectedExecutor);
-            ExecutorManager.this.commonMetrics.markDispatchSuccess();
-          } catch (final ExecutorManagerException e) {
-            ExecutorManager.this.commonMetrics.markDispatchFail();
-            logger.warn(String.format(
-                "Executor %s responded with exception for exec: %d",
-                selectedExecutor, exflow.getExecutionId()), e);
-            handleDispatchExceptionCase(reference, exflow, selectedExecutor,
-                availableExecutors);
+        while (true) {
+          String giveUpReason = checkGiveUpDispatching(reference, remainingExecutors);
+          if (giveUpReason != null) {
+            logger.error("Failed to dispatch queued execution " + exflow.getId() + " because "
+                + giveUpReason);
+            finalizeFlows(exflow);
+            // FAILED TOO MANY TIMES - exit
+            break;
+          } else {
+            final Executor selectedExecutor = selectExecutor(exflow, remainingExecutors);
+            if (selectedExecutor == null) {
+              ExecutorManager.this.commonMetrics.markDispatchFail();
+              handleNoExecutorSelectedCase(reference, exflow);
+              // RE-QUEUED - exit
+              break;
+            } else {
+              try {
+                dispatch(reference, exflow, selectedExecutor);
+                ExecutorManager.this.commonMetrics.markDispatchSuccess();
+                // SUCCESS - exit
+                break;
+              } catch (final ExecutorManagerException e) {
+                logFailedDispatchAttempt(reference, exflow, selectedExecutor, e);
+                ExecutorManager.this.commonMetrics.markDispatchFail();
+                reference.setNumErrors(reference.getNumErrors() + 1);
+                // FAILED ATTEMPT - try other executors except selectedExecutor
+                remainingExecutors.remove(selectedExecutor);
+              }
+            }
           }
-        } else {
-          ExecutorManager.this.commonMetrics.markDispatchFail();
-          handleNoExecutorSelectedCase(reference, exflow);
         }
       }
     }
 
+    private String checkGiveUpDispatching(ExecutionReference reference,
+        Set<Executor> remainingExecutors) {
+      if (reference.getNumErrors() >= this.maxDispatchingErrors) {
+        return "reached " + ConfigurationKeys.MAX_DISPATCHING_ERRORS_PERMITTED
+            + " (tried " + reference.getNumErrors() + " executors)";
```

Comment:
tried ** executions?

Reply:
"Executors" is correct currently, because each available executor is only tried once at most – if all have been tried, give up. Also, this loop is about dispatching a single execution.